### ISSUE TITLE
Multiple CRUD operations per query.

### ISSUE THREAD

**Issue by dmill-bz** [spider/spider#81]
This is an issue to discuss the implementation of CRU and RD scenarios.

Here are some scenarios we need to cover:
1. Create (C) two vertices and an edge between them
2. Find (R) existing vertices and create (C) an edge between them
3. Find (R) existing vertices and create (C) an edge between them then update that edge with properties (U)
4. Create (C) elements and update (U) them with data. 
5. Find (R) and delete (D) elements (should already work but leaving here incase we decide to make it work in a similar fashion to the above)

Lets try and suggest ways of writing the _declarative_ queries for each case:
## 1

``` php
(new BaseBuilder())->create([   
   [TYPE => VERTEX, LABEL => 'person', "name" => 'what'],
   [TYPE => VERTEX, LABEL => 'person', "name" => 'ever'],
   [
      TYPE => EDGE,
      LABEL => 'friend',
      INV => (new BaseBuilder())->create([[TYPE => VERTEX, LABEL => 'person', "name" => 'michael']]),
      OUTV => (new BaseBuilder())->create([[TYPE => VERTEX, LABEL => 'person', "name" => 'dylan']])
   ]
]);
```

I added some extra vertices to show how we could possibly batch
## 2

``` php
(new BaseBuilder())->create([
   [
      TYPE => EDGE,
      LABEL => 'friend',
      INV => (new BaseBuilder())->select()->where(["name", EQUALS, "michael", AND]),
      OUTV =>  (new BaseBuilder())->select()->where(["name", EQUALS, "dylan", AND])
   ]
]);
```
## 3

``` php
(new BaseBuilder())->create([
   [
      TYPE => EDGE,
      LABEL => 'friend',
      INV => (new BaseBuilder())->select()->where(["name", EQUALS, "michael", AND]),
      OUTV =>  (new BaseBuilder())->select()->where(["name", EQUALS, "dylan", AND])
   ]
])
->update()->withData(['weight' => 15]);
```

What to do here if we batch created? The above would apply the data to all of them? How to apply the data selectively?
## 4

Same as above really. With all the same questions.
## 5

Currently we use:

``` php
(new BaseBuilder)->delete()->where(['name', EQUAL, 'dylan', AND]);
```

Don't think this needs to change.

What are your suggestions?

**Comment by electricjones** [comment]
This is basically exactly what I was thinking. A few thoughts:

## Data and fields

I would prefer to use the term 'data' consistently. So, for **1**:

``` php
(new BaseBuilder())->create([   
   [TYPE => VERTEX, LABEL => 'person', "data" => ["name" => 'what']],
   [TYPE => VERTEX, LABEL => 'person', "data" => ["name" => 'ever']],
   [
      TYPE => EDGE,
      LABEL => 'friend',
      INV => (new BaseBuilder())->create([[TYPE => VERTEX, LABEL => 'person', "name" => 'michael']]),
      OUTV => (new BaseBuilder())->create([[TYPE => VERTEX, LABEL => 'person', "name" => 'dylan']])
   ]
]);
```

and so on.

## Batches

For _3_ and _4_, what about

``` php
(new BaseBuilder())->create([
   [
      TYPE => EDGE,
      LABEL => 'friend',
      INV => [(new BaseBuilder())->select()->where(["name", EQUALS, "michael", AND]), $data]
      OUTV =>  [(new BaseBuilder())->select()->where(["name", EQUALS, "dylan", AND]), $data]
   ]
]);
```

## Variables

To simplify matters, I would like to introduce commands as variables, either here or in the `Builder`. Using an example from above:

``` php
$builder = new BaseBuilder();
$builder->create([[TYPE => VERTEX, LABEL => 'person', "name" => 'michael']])->set('michael');
$builder->create([[TYPE => VERTEX, LABEL => 'person', "name" => 'dylan']])->set('dylan');

$builder->create([   
   [TYPE => VERTEX, LABEL => 'person', "name" => 'what'],
   [TYPE => VERTEX, LABEL => 'person', "name" => 'ever'],
   [
      TYPE => EDGE,
      LABEL => 'friend',
      INV => $builder->get('michael'),
      OUTV => $builder->get('dylan'),
   ]
]);
```

This workflow will change our `CommandBag` a bit, but I'd like to place the bulk of the work on the processors. The easiest thing to do is to pass command bags that have nested command bags. And/or allow for an array of command bags to be processed and turned into a batch. In Orient, this shouldn't be difficult. I don't believe it would in cypher.

**Comment by dmill-bz** [comment]
To be honest I don't really like the separation of logic that 'data' implies. After all in 

``` php
[TYPE => VERTEX, LABEL => 'person', "name" => 'what']
```

All of these elements are properties (although some are meta properties : `const`). Separating some into 'data' doesn't make much sense.
I'm actually wondering if `withData()` should even be a thing. Perhaps just place changes directly into `update()` or `create()` instead?

---

Thinking about this again there are two families to the CRUD model. Those that "return" elements (`select`, `create`), and those that modify them (`delete`, `update`). This is in fact the beauty of declarative languages. If we want to be consistent, each query containing the later should have one of the former.

Meaning that the following query is incorrect:

``` php
(new BaseBuilder)->delete()->where(['name', EQUAL, 'dylan', AND]);
```

And should instead be:

``` php
(new BaseBuilder)->select()->where(['name', EQUAL, 'dylan', AND])->delete();
```

You could also imagine : 

``` php
(new BaseBuilder)->create([[TYPE => VERTEX, LABEL => "person", "name" => "dylan"]])->delete();
```

Or even:

``` php
(new BaseBuilder)->create([
   [TYPE => VERTEX, LABEL => "person", "name" => "dylan"],
   [TYPE => VERTEX, LABEL => "person", "name" => "michael"],
])->where(['name', EQUAL, 'dylan', AND])->delete();
```

That would only delete dylan.
This implies changes to the `Bag` and `Processors` but it's for the best.

## 3

With this a CRU query would look something like this for separate updating : 

``` php
(new BaseBuilder())->create([  
   [TYPE => VERTEX, LABEL => 'person', "name" => 'what'],
   [TYPE => VERTEX, LABEL => 'person', "name" => 'ever'],
   [
      TYPE => EDGE,
      LABEL => 'friend',
      INV => (new BaseBuilder())->select()->where(["name", EQUALS, "michael", AND]),
      OUTV =>  (new BaseBuilder())->select()->where(["name", EQUALS, "dylan", AND]),
      "weight" => 23
   ]
]);
```

Here only the edge receive the weight property.
In the event of joint updating : 

``` php
(new BaseBuilder())->create([  
   [TYPE => VERTEX, LABEL => 'person', "name" => 'what'],
   [TYPE => VERTEX, LABEL => 'person', "name" => 'ever'],
   [
      TYPE => EDGE,
      LABEL => 'friend',
      INV => (new BaseBuilder())->select()->where(["name", EQUALS, "michael", AND]),
      OUTV =>  (new BaseBuilder())->select()->where(["name", EQUALS, "dylan", AND])
   ]
])
->update(['weight' => 15]);
```

Here 'what' and 'ever' vertices as well as the edge get the 'weight' property.

And we could also update the inner selects like this:

``` php
(new BaseBuilder())->create([
   [TYPE => VERTEX, LABEL => 'person', "name" => 'what'],
   [TYPE => VERTEX, LABEL => 'person', "name" => 'ever'],
   [
      TYPE => EDGE,
      LABEL => 'friend',
      INV => (new BaseBuilder())->select()->where(["name", EQUALS, "michael", AND])->update(['weight'=>15]),
      OUTV =>  (new BaseBuilder())->select()->where(["name", EQUALS, "dylan", AND])->update(['weight'=>25])
   ]
]);
```

This feels very consistent and allows for all types of operations. 

---

When it comes to `get()` and `set()` shortcuts I'm not sure it makes entire sense to have these on the builder level. First off `get("michael")` doesn't give any information about the field, and `get("name", "michael")` is just the same as a `Builder::where()` so it kind of defeats the purpose. A `set()` with all required fields would probably be too close to an `update()` as well. I say we wait this one out for now.

On the same line of enquiry, `createEdge()` and other methods of the like shouldn't be in the `Builder`. The `Builder` should essentially be CRUD operations (with support for relationships in R) I don't think it's the right level.

The main reason being transaction awareness (that the builder doesn't have). A simple operation like `createEdge()` might be a single query in Orient/neo/gremlin. But might be multiple queries (even if batched) in MYSQL (if/when we make a MYSQL implementation). In this case you might need to update 2 tables and you want the operation to be rollbacked if the second update fails. That's not something the Builder can do. The Model layer would probably be best.

Food for thought.

**Comment by electricjones** [comment]
**First**, I see what you mean about data, and I agree.

**Second**, I wasn't suggesting a `createEdge()` or anything like that being on the `BaseBuilder`, but I _would_ like to see them on the `Builder`, the reason being that not everyone uses models, and definitely not for every operation. All the good ORMs (doctrine, eloquent, propel) that I've seen offer a fully-featured query builder, which is what `Builder` is meant to be.

That said, they do NOT belong on `BaseBuilder` which only performs single or multiple CRUD operations and manages edges. No complex traversals. No multiple step operations that are not explicitly defined.

---

**Third**, I must not have been clear. I intended `get()` and `set()` to save instances of a command bag under that alias only. They in no way affect the actual query. I realize that `get()` currently exists. I would be proposing to change its function entirely.

I find it best to illustrate this first with psudo-sql:

```
CREATE EDGE is_younger_than FROM (SELECT FROM person WHERE name = 'michael') TO (SELECT FROM person WHERE age > 30) DIRECTION OUT SET by_years = 2
```

This creates an outgoing edge from any person named michael to all persons who are older than 30 and sets the by_years property on the edge to 2. Granted, I can't see why you would do this, but the principle illustrates what I mean.

Now, the same script with the introduction of variables, looks like an orient sql batch

```
LET michaels = SELECT FROM person WHERE name = 'michael'
LET targets = SELECT FROM person WHERE age > 30
LET edges = CREATE EDGE is_younger_than FROM $michaels TO $targets DIRECTION OUT SET by_years= 2
commit
return $michaels
```

Ignoring the commit and return for now, I would propose that `BaseBuilder` could do the same:

``` php
$builder = new BaseBuilder();
$builder->select()->from('person')->where(['name', EQUAL, 'michael', AND])->set('michaels');
$builder->select()->from('person')->where(['age', GT, 30, AND])->set('targets');
$builder->create([
      TYPE => EDGE,
      LABEL => 'is_younger_than',
      INV => $builder->get('targets'),
      OUTV =>  $builder->get('michaels'),
      'by_years' => 2
   ]
]);
```

And then, `$builder->getBag()` would produce a single `CommandBag` for creating the edges with two embedded `CommandBags` for selecting targets and michaels. Which the Processor would simply have to parse into a batch or transaction script.

---

**In regards to**

> Thinking about this again there are two families to the CRUD model. Those that "return" elements (select, create), and those that modify them (delete, update). This is in fact the beauty of declarative languages. If we want to be consistent, each query containing the later should have one of the former.

I can see the value of that, so long as that is abstractable in a `Builder`. `$builder->delete->where()` is just so common in ORMs, so readable, and so simple that its hard for me to go against traditional wisdom here. I just feel like `$builder->select()->where()->delete()` can become too convoluted.

---

**Last**, we keep using `create()`. Is that somehow different than the current `insert()` or do we want to modify `insert()` to work as the above? Or, maybe rename `insert()` to `create()`. 

I would probably be in favor of the last one and refactoring `BaseBuilder` be only have 
- `create()`
- `retrieve()`
- `update()`
- `delete()`, 
- `where()`, 
- `projections()`, 
- `limit()`
- `groupBy()`
- `orderBy()`
- `tree()`
- `path()`
  and then the utilities like `getBag()`, etc

Removing: `type()`, `data()` maybe.
In order to handle signature conflicts, we may even want to prefix most of these with "_" or "internal"  or "base" or something. Those that are sweetened in `Builder`.

**Comment by electricjones** [comment]
And my thoughts appreciate the yummy food. Here's some steak for your thoughts :smile:

**Comment by smolinari** [comment]
Hmm...reading your post above Chris and how `set` is used in that example, I think I made a mistake in my PR with set and changing it to getSet. As it seems, it is a setter, in a sense. It might even sound better as `setTo`?

I am going to be a fifth wheel here again for sure. But, could the CRU and similar queries look like this? Or rather, the question in my mind is, are `TYPE` and `LABEL` always a given?

```
(new BaseBuilder())->create([
   [VERTEX => 'person', "name" => 'what'],
   [VERTEX => 'person', "name" => 'ever'],
   [EDGE => 'friend',
      INV => (new BaseBuilder())->select()->where(["name", EQUALS, "michael", AND])->update(['weight'=>15]),
      OUTV =>  (new BaseBuilder())->select()->where(["name", EQUALS, "dylan", AND])->update(['weight'=>25])
   ]
]);
```

And for this example. Not sure this makes any sense, but:

```
$builder = new BaseBuilder();
$builder->begin();
$builder->setTo('michaels')->select()->from('person')->where(['name', EQUALS, 'michael']);
$builder->setTo('tartgets')->select()->from('person')->where(['age', GT, 30]);
$builder->create([EDGE => 'is_younger_than',
      INV => $builder->get('targets'),
      OUTV =>  $builder->get('michaels'),
      'by_years' => 2
   ]
]);
$builder->commit();
```

I would believe that would have to be a transaction, thus the `begin()` and `commit()`. Or would builder know that on its own?

Oh, and EQUAL should be EQUALS. 

Couldn't we just use "="`? I mean, would I have to write "LTEQUALS" for "<="? 

As you can tell, I am a lazy bastard. LOL!

Scott

**Comment by smolinari** [comment]
WIth this...

```
(new BaseBuilder())->create([
   [VERTEX => 'person', 'name' => 'what'],
   [VERTEX => 'person', 'name' => 'ever'],
   [EDGE => 'friend',
      INV => (new BaseBuilder())->select()->where(["name", EQUALS, "michael", AND])->update(['weight'=>15]),
      OUTV =>  (new BaseBuilder())->select()->where(["name", EQUALS, "dylan", AND])->update(['weight'=>25])
   ]
]);
```

and me being a bigger flat fifth wheel maybe...LOL! Go to this?

```
$builder = new Builder();
$builder->createVertex([
   ['person', 'name' => 'what'],
   ['person', 'name' => 'ever']
]);
$builder->createEdge([
   [ 'friend', 
     INV => ($builder->select()->where(["name", "=" , "michael"])->update(['weight' => 15]),
     OUTV =>  ($builder->select()->where(["name", "=" , "dylan"])->update(['weight' => 25])
   ]
]);
```

and getting even wilder, have different functions for plural or singular activities, to simplify the singulars. 

```
$builder = new Builder();
$builder->createVertexes([
   ['person', 'name' => 'what'],
   ['person', 'name' => 'ever']
]);
$builder->createEdge('friend')
             ->inv($builder->select()->where(["name", "=" , "michael"])->update(['weight' => 15]))
             ->outv($builder->select()->where(["name", "=" , "dylan"])->update(['weight' => 25]));

$builder->createEdges([
  ['friend',
      INV => ($builder->select()->where(["name", "=" , "michael"])->update(['weight' => 15]),
      OUTV => ($builder->select()->where(["name", "=" , "dylan"])->update(['weight' => 25]);],
  ['foe',
      INV => ($builder->select()->where(["name", "=" , "bob"])->update(['weight' => 2])),
      OUTV => ($builder->select()->where(["name", "=" , "jane"])->update(['weight' => 3]))]
]);
```

????

Edit: though, thinking about the plural method names, it seems like overkill. Hmm.....

Scott

**Comment by dmill-bz** [comment]
In regards to **Second**. The problem here is that these operations will need to be encapsulated in a transaction. Like I was saying, creating an edge can be a multiple query deal (even if they can be batched into a single request). It's very important that the queries be rollbacked if one fails. A corrupt graph database is absolute hell do deal with :( . `Query`, `Builder` and `BaseBuilder` have no transaction awareness. They're simply string manipulation libraries (even if query is aware of the Driver).

As a general rule `Query`, `Builder`, `BaseBuilder` should only produce failure safe queries. Meaning that any query generated if it were to fail should fail as a whole. If we use batches we want to make sure that any mutations that occur would rollback automatically if one of the later operations failed. (this will be DB specific)

Another reason why Builder isn't such a good place for this type of abstraction is that if we produce a list of similar abstractions a lot of them clearly can't sit within `Builder`. Having some in one place and others in another place is a little sloppy.
- `createEdge()` With some DBs this could be a twofold operation. Modifying the references in two tables for example, you would need the first reference to revert back if the second update fails. (though potentially all scenarios could play out in single queries which would make this eligible for `Builder`)
- `updateEdge()` The naming here is very vague but the idea is that some update operations (like changing the label) involves cloning the existing edge, deleting it, then adding the clone with a different label. If the adding fails (because of say, an existing unique index) then you want the delete to be rolled back.
- `changeOutV`/`changeInV` Changing an edge's outV or inV is the same as above. You need to remove the edge and re-insert it. This is a great abstraction to have though. Very handy.
- `updateVertex()` Again this is a general name but it's similar to the `updateEdge()`, the vertex needs to be removed and added back in with a different label. It's actually even heavier as all the adjacent edges need to be removed and added back in again.

Basically none of these operations can simply be added to `Builder` or `Query` without forming some integrity hasard. Or in the case of createEdge.. well it would be alone in `Builder` while the others reside elsewhere (and that's if we're certain it won't produce issues with some DBs).

So yes it would be cool to have this on the Builder level, but it doesn't seem realistically feasible. Anyone not using Spider Models are probably using their own models (or whatever models come with the framework they're using) and will have to implement the abstractions themselves. Which I think is legitimate.
There might be another way of providing this level of abstraction in a transaction aware way that isn't a model. Worth thinking about.

For **Third** thanks for clearing that out. I didn't understand that at all. I see what you mean now. So essentially its an alternative way to how we originally wrote the query. Instead of nesting the queries manually the builder does it? Does this mean we write the `BaseBuilder` for nesting and provide this as a sugar ?

**Last**, I'm down with renaming `insert()` to `create()` and I also think we should make the naming consistent. I tend to like the `internal` prefix better but I'm not against `_` either.

@smolinari `TYPE` and `LABEL` aren't always a given. Actually a lot of existing systems and DBs don't support labels for vertices (of course these are all mostly older versions but a lot of corporate scenarios take forever to update).
Also in some new implementations you might not always have a vertex label (not necessary in multiple cases). Some people might prefer to use custom indexes they have more control over rather than labels.
Cases in which `TYPE` isn't a given might be a bit of a stretch though.

For the `EQUAL`/`EQUALS`/`=` cases. Only `BaseBuilder` uses the `const`,  `Builder` provides sugar that allows you to use `=` etc. :)

A lot of the "lazy" work of providing an easy non-verbose query builder will come once we have the `BaseBuilder` set. It'll all be in the `Builder` implementation.

**Comment by smolinari** [comment]
Ah, ok. To be honest, I was wondering why BaseBuilder was being used the whole time in the discussion. So, I'll go back under my rock again. :smile: 

Scott

**Comment by electricjones** [comment]
Alright, I think we're getting somewhere.

First @smolinari, your PR was fine. I was proposing a further change. ATM, there is no functionality like `set()` and `get()` that I described.

---

As far as **failure safe operations at one of the `Builder` levels**. I had assumed that the three drivers we are using offered a way (without implicit transactions) to do batch operations in an atomic way. In OrientSQL batch (which all operations to through now) they are atomic. That is, if you are selecting two vertices and creating an edge and any one part of that stack fails, all parts are rolled back.

I had assumed this to be implicit in Gremlin and Cypher. Are you telling me its not? I didn't mind `createEdge()` et all at the `Builder` level because I knew that the Orient Processor could produce an atomic-safe script to execute these 2 or three steps. If this isn't the case with the other two drivers, then I see the problem.

My thought was that _any_ operations inside a single `CommandBag`, including all of its nested `CommandBag`s would be atomic. The driver transactions would be used by users who want to do even more operations atomically.

---

On **nested queries and variables**, yes this is an alternative. I don't know that I'd call it sugar, but its similar. Its a way to save queries and reuse them at runtime.

Say you have a complicated query with a lot of wheres that selects all users with a specific favorite color, place, yadda, yadda, yadda. You can go through this query, `set()` (or maybe `save()`) it with an alias, and then nest it into another operation later. Even nest it into several operations.

---

Let's stick with `internal` prefix for `BaseBuidler`, and move `type` to `Builder`. It's just a where statement.

**Comment by dmill-bz** [comment]
This all sounds good.
I think it's safe to say that Gremlin and Neo can both provide a failure safe edge creation. I was thinking more along the lines of a mongo DB implementations for example (or any of the big table backends). Something that isn't a qualified graph DB but could still be used with spider.

**Comment by electricjones** [comment]
It has always been my intention to write drivers for MySQL and Mongo at some point. In that odd edge case, I think it would be up to the driver and processor to play nicely.

For example, the processor would create a custom script like:

```
begin CustomScript
0 => // OPERATION ONE
1 => // OPERATION TWO
2 => // OPERATION THREE
end CustomScript
```

And the driver would have to do a simple regex search of scripts. If it encounters a custom script, then it starts a transaction, cycles through the operations, and commits. Its all atomic. The implementation may not be so easy, but I think its doable. We could even add an `AbstractProcessor` and some methods to `AbstractDriver` to help with this when the time comes.

---

Please edit this post to add the tasks we are moving forward with.

So, we are settling on edge management for `BaseBuilder`, ignoring `Builder` and `Models` for the moment:
- [x] Move `type()` and `data()` out of `BaseBuilder` and rename `insert()` to `create()`
- [x] Prefix all methods with `internal` inside `BaseBuilder`
- [ ] Allow for nested queries via injecting instance of `BaseBuilder` or `CommandBag`

``` php
(new BaseBuilder())->create([
   [VERTEX => 'person', "name" => 'what'],
   [VERTEX => 'person', "name" => 'ever'],
   [EDGE => 'friend',
      INV => (new BaseBuilder())->select()->where(["name", EQUALS, "michael", AND])->update(['weight'=>15]),
      OUTV =>  (new BaseBuilder())->select()->where(["name", EQUALS, "dylan", AND])->update(['weight'=>25])
   ]
]);
```
- [x] No to `createEdge()`, etc. Let's just use the CRUD methods we have for all types.

**Comment by dmill-bz** [comment]
@smolinari Don't worry, we really appreciate your feedback. And seeing as how the discussion is going we will probably implement a variation of your vertex and edge creation methods. So don't stop ;)

@chrismichaels84 
Ok I see where you're going with the edge cases (no pun intended) around the edge creation scenarios. This feels a little hacky but it might be the only way to get it to work.

In the upcoming tasks:
Should we just remove `data()` & `withData()` in favor of populating `Bag::$update` and `bag::$create` on `BaseBuilder::internalUpdate($data)` and `BaseBuilder::internalCreate($data)`?
Also, should we leave `set()`and `get()` to later?

**Comment by electricjones** [comment]
Yes to removing `data()` and `withData()` at least in the `BaseBuilder` which is my only concern for now.

I would like `set()` and `get()` on the `BaseBuilder`, but let's tackle things one step at a time. Yes.

**Comment by dmill-bz** [comment]
Oh and also, what of:

``` php
(new BaseBuilder)->select()->where(['name', EQUAL, 'dylan', AND])->delete();
```

?
It diverges a bit from what you're commonly used to seeing but it flows with the logic of the rest of `Builder`.
Maybe this can be the `BaseBuilder` behavior. Builder would just translate `delete()` to the combination of `select()->delete()` in `BaseBuilder`?

**Comment by electricjones** [comment]
For `BaseBuilder` I do like this. It is explicit, which seems to be the name of the game for `BaseBuilder`. We can discuss the way to sweeten everything in `Builder` once we have this squared away.

**Comment by electricjones** [comment]
So, it seems like we've settled on a workflow for multiple CRUD operations inside `BaseBuilder` and I like it because it doesn't introduce much new stuff into `BaseBuilder` itself.

Most of the work will fall into the `Processors` and maybe a little into the `Bag`

Is any of this affected by our discussion #53? `LinkBag`?

**Comment by dmill-bz** [comment]
I don't think so. Even if there was something it's probably best to move forward with this separately and then change it when we PR for #53

**Comment by electricjones** [comment]
So, the last question I have is if we are going with `$select->from()->where()->delete()` what would the `CommandBag` look like If a single `CommandBag` can hold multiple commands?

The way I see it, there are a few options.
1. `BaseBuilder` can hold a `CommandBag` or an array of `CommandBags` which are then executing in order in an atomic way. The problem with this is knowing when and how to use the first command in the second. This would be the problem with `$select()->delete()`
2. Introduce a `CommandBagManager` or something that `BaseBuilder` can manipulate by initializing and populating single `CommandBag`s.
3. Enable `CommandBag` to separate multiple commands, somehow.

I would really like a 4th, simpler option.

To be clear, this is not a concern about _nested_ commands. That seems pretty straightforward.

**Comment by smolinari** [comment]
`$select->from()->where()->delete()`

I feel an "action method" needs to be where it makes the most sense for reading and should be as close as possible to what is common in other ORMs currently and to SQL, when possible. So I like what Dylan suggest in the first post. 

`(new BaseBuilder)->delete()->where(['name', EQUAL, 'dylan']);`

Scott

**Comment by dmill-bz** [comment]
Putting `set()`and `get()` aside, I think all we need to add is `Bag::$create`, `Bag::$delete`, and `Bag::$update`. All the rest of the changes are in the processor.

**Comment by dmill-bz** [comment]
Going to elaborate on the above. This also makes `Bag::$command` pointless. We can just check whether or not the following are set:
- `Bag::$projections` : not null means select scenario
- `Bag::$update` : not null means update scenario
- `Bag::$create` : not null means create scenario
- `Bag::$delete` : not null means delete scenario

And a combination of the above is obviously possible.

**Comment by dmill-bz** [comment]
Making multiple posts as I go but it also means we should check that there is at least one select or create scenario. (that's the only check we need to make to see if a `Bag` is viable.)

**Comment by electricjones** [comment]
Alright, I'm on board with that. So the processors also validate bags? Sounds like its time for an `AbstractProcessor`.

I can take a whack at the `BaseBuilder` tomorrow. It should be simple enough to:
- [x] normalize the method names (with baseX) and move `type()` and `data()` into `Builder`
- [ ] Update `Bag` with the above
- [ ] Create a an `AbstractProcessor` and initialize a `validate` method.

I want to get this passed before we start to tackle the processors. It will go much more smoothly if we have a more-less set in stone builder and bag.

**Comment by electricjones** [comment]
If you want to start looking at a `GremlinProcessor`, now may be a good time to get that ball rolling.

**Comment by electricjones** [comment]
Or, you can begin work on #53, but that may be easier after these changes take affect. (apparently multiple comments is the theme for the day)

**Comment by dmill-bz** [comment]
Given how Gremlin works it's probably best for me to tackle that one once all the pieces are in place. I'll take a look at #53, it might be doable without these changes.

**Comment by electricjones** [comment]
Great

**Comment by electricjones** [comment]
Alright, I don't know where my head was, but I wasn't wrapping by brain around what you were saying until this afternoon and I'm totally onboard. I went through several gremlin and cypher tutorials and see the need and use cases more clearly. The need just isn't as prevalent in Orient, but I need to branch out anyway. Call it a learning experience.

In any case, I am finishing up a PR that does several things:
- [ ] Cleans up `Bag` by
  - [ ] removing `command` and some other old code
  - [ ] adding `create`, `update`, and `delete`
  - [ ] renaming `projections` to `retrieve` (for consistency)
- [ ] Refactors `BaseBuilder` to manipulate the new Bag
- [ ] Adds new tests for multiple operation CRUD

I have run into a couple questions that I wanted your opinion on. My opinions follow.

## 1. Target

I have removed `target` from `Bag`. I feel like it belongs as a `where` constraint, which is what it really is in every datastore except Orient and Arrango. That said, there does need to be a default. Even if the user never specifies `->where()`, the target should default to vertex:
`where => [[TYPE, EQUAL, VERTEX, AND]]`

My question is, does the `Bag` set this as a default, does the `BaseBuilder` add this if it is missing, or does the `Processor` assume this if no other ELEMENT_TYPE constraint is given?

My preference would be for the `Bag` to hold the default so that `BaseBuilder` isn't aware of it and `Processors` are all the same across the board. To do this will probably necessitate a method or two (some setters and getters) inside `Bag` to ensure the defaults are set, but can be overridden.

## Edge Validation

A similar question for validation when creating an edge. An Edge should require an INV and OUTV, right? Where is the exception thrown if missing? Which component validates: `BaseBuilder`, `Processor`, or `Bag`?

My gut instinct says `Processor`. I would like to keep all `Bag` validation in one place. That could either be in an `AbstractProcessor` or, now that I think about it, a `validate` method on `Bag` since `Bag will be aware of what it should be like. For that matter it may be worth adding a`loadDefaults`method to`Bag`to setup the defaults for target, etc. And then a`prepare` method, which simply loads defaults and validates.

## Ordering of operations

If we are truly going for a declarative approach as you illustrated in #85, then I would think the following _would_ be possible:

``` php
(new BaseBuilder)->delete()->where()->limit();
```

Just as much as much as 

``` php
(new BaseBuilder)->select()->where()->limit()->delete();
```

Basically, I don't see why a `select` is _required_ before a delete or update. They aren't for Arrango or Orient. I think either approach should be valid and the processor sort out the best way to write an efficient command.

**Comment by dmill-bz** [comment]
**Target**
I agree that this is totally a where clause. Personally I feel that the `Processor` shouldn't care for it, and I'm torn between placing the logic in the `Bag` and `Builder`. It's anyone's guess at this point, anyways it shouldn't be hard to move around if we decide to go with the other option.

**Edge validation**
In general I think the DB can complain about this kind of stuff. Just as if you wrote an incorrect query. 
With that said, in some cases with gremlin you only need either `OUT_V` or `IN_V`. So maybe adding validation for this would be a good way to ensure that DBs can plug&play.

Perhaps putting this in the `Bag` makes sense? That way it validates itself?

Also on the Edge front. Both gremlin and Neo may require relationship logic in the bag for edge insertion to work. Maybe we shouldn't be too hasty in implementing this.

**Ordering of operations**
It was a matter of being consistent. The two queries you show actually shouldn't be interchangeable (though they just happen to "kind of" be because delete doesn't take any arguments)

For instance : 

``` php
(new BaseBuilder)->select("b.*")->where(/* set what b refers to */)->limit()->delete();
```

Doesn't have an equivalent without `select()` defining the object since `delete()` doesn't take any arguments. It's even easier to see how this could complicate things with an update or the likes that already do take arguments.

Of course this doesn't mean we can't provide sugar in `Builder` that would allow for : 

``` php
(new BaseBuilder)->delete("b.*")->where(/* set what b refers to */)->limit();
```

But this would be a one off and behave inconsistently with the rest of the builder. Suddenly `delete` is in charge of defining the objects we want to retrieve... A little weird. Also if we ever decide to give delete params, we're in a world of trouble.

I understand that it might seem more in line with SQL, but SQL was already a little inconsistent to start with.

**Comment by electricjones** [comment]
I don't like the idea of giving `delete` parameters, but I _do_ think it is more intuitive to give the option of a single operation delete and update. We can do this with sugar in the `Builder`. For now, let's shelve that.

I am going to put validation and defaults in the `Bag` for now and do everything I can to make `BaseBuilder` and `Bag` as simple as possible.
